### PR TITLE
[v4r1] Change realization oauth2 from js library to dirac service(associated with OAuthDIRAC) #315

### PR DESCRIPTION
**Pull request by TaykYoku**
BEGINRELEASE

*Lib/Conf.py
CHANGE: Added configuration helpers for authorisation handler

*Lib/WebHandler.py
FIX: Fixed some bugs in oauth2 authorisation logic

*WebApp/handler/AuthenticationHandler.py
CHANGE: Changed logic to work with oauth2 service

*WebApp/static/core/js/views/desktop/TaskBar.js
CHANGE: Changed logic to work with oauth2 service

*WebApp/static/oidc
FIX: Removed

*WebApp/template/root.tpl
FIX: Removed OIDC js library

ENDRELEASE

### CODE REVIEW

**Comment by TaykYoku** [comment]
OK for merging?

File: Lib/WebHandler.py [hunk 1]
```diff
@@ -166,39 +172,49 @@ def __processCredentials(self):
...
-        if result['OK']:
-          self.__credDict['username'] = result['Value']
-        result = getDNForUsername(self.__credDict['username'])
-        if result['OK']:
-          self.__credDict['validDN'] = True
-          self.__credDict['DN'] = result['Value'][0]
-        result = getCAForUsername(self.__credDict['username'])
+      if self.get_secure_cookie("StateAuth"):
+        state = self.get_secure_cookie("StateAuth")
+        result = oauth.getUsrnameForState(state) if oauth else S_ERROR('OAuthDIRAC extantion is not enable')
```
Comment (fstagni) [inline Lib/WebHandler.py]:
```suggestion
        result = oauth.getUsrnameForState(state) if oauth else S_ERROR('OAuthDIRAC extension is not enable')
```

File: Lib/WebHandler.py [hunk 2]
```diff
@@ -1,28 +1,33 @@
...
 
-from WebAppDIRAC.Lib.SessionData import SessionData
 from WebAppDIRAC.Lib import Conf
-from DIRAC.ConfigurationSystem.Client.Helpers.Registry import getUsernameForID, getDNForUsername, getCAForUsername
+from WebAppDIRAC.Lib.SessionData import SessionData
+
+try:
+  from OAuthDIRAC.FrameworkSystem.Client.OAuthManagerClient import OAuthManagerClient
+  oauth = OAuthManagerClient()
+except Exception as e:
```
Comment (fstagni) [inline Lib/WebHandler.py]:
```
except ImportError as e:
  <some error>
except BaseException:
  <some other error> + log.exception()
```

File: Core/HandlerMgr.py [hunk 3]
```diff
@@ -1,42 +1,45 @@
...
-    """
-    Get lists of paths for all installed and enabled extensions
+    """ Get lists of paths for all installed and enabled extensions
     """
     pathList = []
     for extName in CSGlobals.getCSExtensions():
       if extName.rfind("DIRAC") != len(extName) - 5:
         extName = "%sDIRAC" % extName
-      if extName == "WebAppDIRAC":
-        continue
```
Comment (zmathe) [inline Core/HandlerMgr.py]: Why you removed this? This part the code is handling the extensions such as LHCbWebAppDIRAC.
Are you sure it will work without this lines?

File: Core/HandlerMgr.py [hunk 4]
```diff
@@ -55,17 +58,19 @@ def __calculateRoutes(self):
...
-    result = ol.getObjects(origin, parentClass=WebHandler, recurse=True)
-    if not result['OK']:
-      return result
-    self.__handlers = result['Value']
+    self.__handlers = collections.OrderedDict()
+    for origin in self.__sysService:
+      result = ol.getObjects(origin, parentClass=WebHandler, recurse=True)
+      if not result['OK']:
+        return result
+      self.__handlers = collections.OrderedDict(list(self.__handlers.items()) + list(result['Value'].items()))
```
Comment (zmathe) [inline Core/HandlerMgr.py]: You recreating the self.__handlers few times. Could you please add the new handlers to the orderedDict instead of recreate len(self.__sysService) time?

File: Core/TemplateLoader.py [hunk 5]
```diff
@@ -10,10 +11,9 @@ def __init__(self, pathList, **kwargs):
     self.pathList = pathList
 
   def resolve_path(self, name, parent_path=None):
-    if parent_path and not parent_path.startswith("<") and \
-            not parent_path.startswith("/") and \
-            not name.startswith("/"):
-      name = os.path.join(os.path.dirname(parent_path), name)
+    if parent_path and not parent_path.startswith("<"):
+      if not parent_path.startswith("/") and not name.startswith("/"):
+        name = os.path.join(os.path.dirname(parent_path), name)
```
Comment (zmathe) [inline Core/TemplateLoader.py:16]: Why this change is needed?  The logic has not changed. No?

File: Lib/Conf.py [hunk 6]
```diff
@@ -167,7 +170,8 @@ def SSLProrocol():
 
 
 def getStaticDirs():
-  return getCSValue("StaticDirs", [])
+  staticDirs = getCSValue("DefaultStaticDirs") or ['defaults', 'demo']
```
Comment (zmathe) [inline Lib/Conf.py]: Why we need DefaultStaticDirs and StaticDirs in CS? What is the different?

File: Lib/WebHandler.py [hunk 7]
```diff
@@ -166,39 +172,49 @@ def __processCredentials(self):
...
-        if result['OK']:
-          self.__credDict['validDN'] = True
-          self.__credDict['DN'] = result['Value'][0]
-        result = getCAForUsername(self.__credDict['username'])
+      if self.get_secure_cookie("StateAuth"):
+        state = self.get_secure_cookie("StateAuth")
+        result = oauth.getUsrnameForState(state) if oauth else S_ERROR('OAuthDIRAC extantion is not enable')
+        if not result['OK']:
+          return result
+        self.set_secure_cookie("StateAuth", result['Value']['state'])
```
Comment (zmathe) [inline Lib/WebHandler.py:183]: probably better: result['Value'].get('state')

File: Lib/WebHandler.py [hunk 8]
```diff
@@ -166,39 +172,49 @@ def __processCredentials(self):
...
+        self.__credDict['username'] = result['Value']['username']
```
Comment (zmathe) [inline Lib/WebHandler.py:184]: result['Value'].get('username') ?

File: Lib/WebHandler.py [hunk 9]
```diff
@@ -1,28 +1,33 @@
-
-from DIRAC import gLogger
-from DIRAC.Core.Security.X509Chain import X509Chain
+from DIRAC import gLogger, S_OK, S_ERROR
 from DIRAC.Core.Security import Properties
-from DIRAC.Core.DISET.ThreadConfig import ThreadConfig
```
Comment (zmathe) [inline Lib/WebHandler.py]: hm. This is used no? When it is removed? This is used to deal with DISET configuration

File: WebApp/handler/AuthenticationHandler.py [hunk 10]
```diff
@@ -1,79 +1,132 @@
...
+  @asyncGen
+  def web_getAuthNames(self):
+    """ Get list of enable authentication types
+    """
+    self.finish(Conf.getAuthNames())
 
-  # Get information from web.cfg about auth types
   @asyncGen
   def web_getAuthCFG(self):
+    """ Get option from IdP
```
Comment (zmathe) [inline WebApp/handler/AuthenticationHandler.py:48]: You want to give access to the configuration file without any authorization authentication? Is the CS section contains confident information?

File: WebApp/handler/AuthenticationHandler.py [hunk 11]
```diff
@@ -1,79 +1,132 @@
...
-    self.write(current)
+    self.finish(current)
+
+  @asyncGen
+  def web_waitOAuthStatus(self):
+    """ Listen authentication status on OAuthDB
+    """
+    state = str(self.request.arguments["state"][0])
+    typeAuth = str(self.request.arguments["typeauth"][0])
+    gLogger.debug('Read authentication status of "%s" session' % state)
```
Comment (zmathe) [inline WebApp/handler/AuthenticationHandler.py]: glogger.debug("Read authentication status of ", "%s session" % state)

**Review by zmathe — CHANGES_REQUESTED** [review]

File: Core/TemplateLoader.py (hunk 5 above)
Comment (TaykYoku) [reply Core/TemplateLoader.py:16]: This change resolve format check error

File: Lib/WebHandler.py (hunk 9 above)
Comment (fstagni) [reply Lib/WebHandler.py]: It's just moved a few lines below.

Comment (TaykYoku) [reply Lib/WebHandler.py]: All libraries are present, I just changed the order for better readability
If I understand correctly this change has no consequences

File: Lib/Conf.py (hunk 6 above)
Comment (TaykYoku) [reply Lib/Conf.py]: In Core/HandlerMgr.py is added some standard paths for static files:
`68 | statDirectories = ['defaults', 'demo'] + Conf.getStaticDirs()`

But for example this paths don't need for http endpoint of OAuthService. So, is added posibility to configure this default paths.

File: Lib/WebHandler.py (hunk 7 above)
Comment (TaykYoku) [reply Lib/WebHandler.py:183]: Added condition that must check dict

File: Lib/WebHandler.py (hunk 8 above)
Comment (TaykYoku) [reply Lib/WebHandler.py:184]: Added condition that must check dict

File: WebApp/handler/AuthenticationHandler.py (hunk 10 above)
Comment (TaykYoku) [reply WebApp/handler/AuthenticationHandler.py:48]: It contain IdPs some secret information(like as client_id and client_secret), that not returned for this method

File: Lib/WebHandler.py (hunk 2 above)
Comment (TaykYoku) [reply Lib/WebHandler.py]: fixed

File: Core/HandlerMgr.py (hunk 4 above)
Comment (TaykYoku) [reply Core/HandlerMgr.py]: fixed

File: WebApp/handler/AuthenticationHandler.py (hunk 11 above)
Comment (TaykYoku) [reply WebApp/handler/AuthenticationHandler.py]: fixed

File: Core/HandlerMgr.py (hunk 3 above)
Comment (TaykYoku) [reply Core/HandlerMgr.py]: I don't remember actually. I agree that logic here not complete without this lines. Now its fixed.

File: Core/TemplateLoader.py (hunk 5 above)
Comment (zmathe) [reply Core/TemplateLoader.py:16]: ok.

File: Lib/Conf.py (hunk 6 above)
Comment (zmathe) [reply Lib/Conf.py]: but you break the compatibility now. Imagine if someone have the following setup:
DefaultStaticDirs = 'a','b' and also using defaults as well.
In this modification only 'a' and 'b' directories will be served... LHCb is not the case but probably others experiments are using some of the directory. Please keep the backward compatibility.

File: Lib/WebHandler.py (hunk 9 above)
Comment (zmathe) [reply Lib/WebHandler.py]: yes, Sorry for that I have not seen

File: WebApp/static/core/js/views/desktop/TaskBar.js [hunk 12]
```diff
@@ -5,206 +5,230 @@
...
+    
+    var getAuthCFG = function(Auth){
       var req = Ext.Ajax.request({
         url: GLOBAL.BASE_URL + 'Authentication/getAuthCFG',
-        params: {
+        params: { typeauth: Auth },
+        async: false
+      }).responseText
+      res = JSON.parse(req)
+      console.log(res)
```
Comment (zmathe) [inline WebApp/static/core/js/views/desktop/TaskBar.js]: please remove console.log. If you want to add debug message, please use 
 GLOBAL.APP.CF.log("debug",

File: WebApp/static/core/js/views/desktop/TaskBar.js [hunk 13]
```diff
@@ -5,206 +5,230 @@
...
           typeauth: Auth,
-          value: Value
+          state: Value
         },
         async: false
       }).responseText
-      res = JSON.parse(req)
-      if (Object.keys(res).includes('Value')) {
-        res = res.Value
+      console.log(status_req);
```
Comment (zmathe) [inline WebApp/static/core/js/views/desktop/TaskBar.js]: do not use console.log use  GLOBAL.APP.CF.log("debug", instead

File: WebApp/static/core/js/views/desktop/TaskBar.js [hunk 14]
```diff
@@ -5,206 +5,230 @@
...
+      status_json = JSON.parse(status_req);
+      console.log(status_json);
```
Comment (zmathe) [inline WebApp/static/core/js/views/desktop/TaskBar.js]: here as well

File: WebApp/static/core/js/views/desktop/TaskBar.js [hunk 15]
```diff
@@ -5,206 +5,230 @@
...
-                        params: { 
-                          typeauth: name,
-                          value: response.profile
-                        },
-                        success: function() { GLOBAL.APP.CF.alert('Your request was sent.','info')  }
-                      });
-                      this.up('window').close();
-                    }
+    var auth = function(AuthType) {
+      console.log('"' + AuthType + '" was chosen for authorization')
```
Comment (zmathe) [inline WebApp/static/core/js/views/desktop/TaskBar.js]: here as well

File: WebApp/static/core/js/views/desktop/TaskBar.js [hunk 16]
```diff
@@ -5,206 +5,230 @@
...
+      result = getAuthCFG(AuthType)
+      if (!result.OK) { return GLOBAL.APP.CF.alert("Authentication was ended with error: %s", result.Message, 'error') }
+      if (result.Value.providerType) {
+        providerType = result.Value.providerType
+        if (providerType == 'OAuth2') {
+          console.log('OIDC authorization flow')
```
Comment (zmathe) [inline WebApp/static/core/js/views/desktop/TaskBar.js]: here as well

File: WebApp/static/core/js/views/desktop/TaskBar.js [hunk 17]
```diff
@@ -5,206 +5,230 @@
...
-                          value: response.profile
-                        },
-                        success: function() { GLOBAL.APP.CF.alert('Your request was sent.','info')  }
-                      });
-                      this.up('window').close();
-                    }
+    var auth = function(AuthType) {
+      console.log('"' + AuthType + '" was chosen for authorization')
+      result = getAuthCFG(AuthType)
+      if (!result.OK) { return GLOBAL.APP.CF.alert("Authentication was ended with error: %s", result.Message, 'error') }
```
Comment (zmathe) [inline WebApp/static/core/js/views/desktop/TaskBar.js]: new line after {
to be consistent

File: WebApp/static/core/js/views/desktop/TaskBar.js [hunk 18]
```diff
@@ -5,206 +5,230 @@
...
+      if (result.Value.providerType) {
+        providerType = result.Value.providerType
+        if (providerType == 'OAuth2') {
+          console.log('OIDC authorization flow')
+          if (!result.Value.settings) { return GLOBAL.APP.CF.alert('We cannot get authorization URL for you.', 'error') }
```
Comment (zmathe) [inline WebApp/static/core/js/views/desktop/TaskBar.js]: new line after {

File: WebApp/static/core/js/views/desktop/TaskBar.js [hunk 19]
```diff
@@ -5,206 +5,230 @@
...
+          authorizationURL = result.Value.settings.url
+          state = result.Value.settings.state
+          
+          //Open popup
+          console.log('Open authorization URL: "' + authorizationURL + '"')
```
Comment (zmathe) [inline WebApp/static/core/js/views/desktop/TaskBar.js]: no console.log

File: WebApp/static/core/js/views/desktop/TaskBar.js [hunk 20]
```diff
@@ -5,206 +5,230 @@
...
+          var oAuthReqWin = open(authorizationURL, "popupWindow", "hidden=yes,height=570,width=520,scrollbars=yes,status=yes")
+          oAuthReqWin.focus();
+          
+          // Send request to redirect URL about success authorization
+          console.log('Watch when popup window will be close')
```
Comment (zmathe) [inline WebApp/static/core/js/views/desktop/TaskBar.js]: no console.log

File: WebApp/static/core/js/views/desktop/TaskBar.js [hunk 21]
```diff
@@ -5,206 +5,230 @@
...
-          }
-        });
-      });
-      manager.signinPopup().catch(function(error){ GLOBAL.APP.CF.log("error",
-      'error while logging in through the popup'); });
-    } 
+          })(120,'opened');
+        } else { GLOBAL.APP.CF.alert("Authentication method " + providerType + " is not supported." , 'error') }
+      } else {
+        console.log('"' + result.Value + '" is null')
```
Comment (zmathe) [inline WebApp/static/core/js/views/desktop/TaskBar.js]: no console.log

**Review by zmathe — COMMENTED** [review]
Please do not use console.log
and please use the correct code formatting for javascript: https://github.com/DIRACGrid/WebAppDIRAC/blob/integration/extjs-template.xml

File: Lib/Conf.py (hunk 6 above)
Comment (TaykYoku) [reply Lib/Conf.py]: In this case a someone dont want to use defaults ['defaults', 'demo'] and he replaced it on ['a', 'b'] and after that in result we have:
StaticDirs +  ['a', 'b']
If someone want just add to defaults ['defaults', 'demo'] some statics, he have original option "StaticDirs" is this case we have:
StaticDirs +  ['defaults', 'demo']

And with the last commit:
 if someone do not need any static dirs he just set DefaultStaticDirs to 'None' in this case we have:
StaticDirs +  [ ]

Comment (zmathe) [reply Lib/Conf.py]: fine with me

**Comment by TaykYoku** [comment]
OK for merging?

**Comment by fstagni** [comment]
I think it's ready for being merged (but remove the WIP:).

But, for @atsareg : we should create a `rel-v4r0` branch (from what's now the content of the `integration` branch) and then simply merge this one into integration, and this will be part of `v4r1` WebAppDIRAC version.

**Comment by atsareg** [comment]
After merging v4r0 into integration there is a conflict now. Please, resolve it

**Comment by TaykYoku** [comment]
> After merging v4r0 into integration there is a conflict now. Please, resolve it

done
